posts/2023-06-12-streams-in-scala.adoc

```diff
@@ -0,0 +1,214 @@
+
+[[scala-stream]]
+== Scala Stream
+image::../media/2023-06-12-streams-in-scala/scala-stream.png[]
+
+The Stream in Scala is part of the collection hierarchy which extends LinearSeq. They're like views, only the elements that are accessed are computed. Other than this behaviour, a Stream behaves similarly to a List. The elements of stream are lazily computed. In Scala Stream, only the first element are pre-computed but as of Scala 2.13, Stream was replace with LazyList where no element is computed unless requested. LazyList is designed to address the issues with Stream and provides a more predictable evaluation model.
+
+[source,scala]
+----
+// Stream prior to 2.13
+(1 to 10000).toStream // output: Stream(1, <not computed>)
+
+// LazyList >= 2.13
+(1 to 10000).to(LazyList) // output: LazyList(<not computed>)
+----
+=== Transformer Vs Terminator method:
+It is important that we know what these are. Transformer methods are collection methods that convert a given input collection to a new output collection, based on an algorithm you provide to transform the data. This includes methods like map, filter, and reverse. On the other hand Terminator methods are collection methods that perform a final computation on a collection and return a non-collection result, such as a single value or a boolean value. These methods effectively terminate the computation and produce a final output. Examples of terminator methods include methods like fold, reduce, count, and exists.  Be careful with methods that aren’t transformers. Calls to the following strict methods are evaluated immediately and can easily cause java.lang.OutOfMemoryError errors:
+
+[[call-by-name]]
+== Call-by-name ( CBN )
+Whilst you can begin using LazyList collection with the information provided so far, I think it would also be good to have a basic understanding of LazyList Implementation. Call-by-name (also known as pass-by-name) is a parameter evaluation strategy in programming languages where the argument expression is not evaluated before it is passed to a function or method. Instead, the expression is evaluated each time it is referenced within the function or method body. Just for note, the other parameter evaluation strategy is called, Call-by-value ( CBV )
```

Comment:
```suggestion
While you can begin using LazyList collections with the information provided so far, I think it would also be good to have a basic understanding of the LazyList Implementation. Call-by-name (also known as pass-by-name) is a parameter evaluation strategy in programming languages where the argument expression is not evaluated before it is passed to a function or method. Instead, the expression is evaluated each time it is referenced within the function or method body. Just for note, the other parameter evaluation strategy is called, Call-by-value ( CBV )
```